posts/Gunnar/2017-01-31-preventing-leaky-apis-with-jqassistant.adoc

```diff
@@ -0,0 +1,227 @@
+= "Tool Time: Preventing leaky APIs with jQAssistant"
+Gunnar Morling
+:awestruct-tags: [ "Discussions" ]
+:awestruct-layout: blog-post
+---
+
+If you've ever watched the great show "Home Improvement", you'll know that a https://en.wikipedia.org/wiki/List_of_Home_Improvement_characters#Tim_Taylor[fool with a tool] is still a fool.
+At the same time though, the right tool used in the right way can be very effective for solving complex issues.
+
+In this post I'd like to introduce a tool called https://jqassistant.org/[jQAssistant] which I've found very useful for running all sorts of analyses of a project's code base,
+e.g. for preventing the leakage of internal types in the public API of a library.
+This is planned to be the first post in a blog series on developer-centric tools we've come to value when working on the different libraries of the Hibernate family.
+
+== Keeping your API clean
+
+When providing a library it is an established best practice to clearly distinguish between those parts of the code base intended to be accessed by users (the API of the library) and those parts not meant for outside access.
+
+Having a clearly defined API helps to reduce complexity for the user
+(they only need to learn and understand the API classes but not all the implementation details),
+while giving the library authors freedom to refactor and alter implementation classes as they deem it necessary.
+Usually, the separation of API and implementation types is achieved by using specific package names, with all the non-public parts being located under a package named `internal`, `impl` or similar.
+
+One thing you have to be very careful about though is to not leak any internal types in the API.
+E.g. a method definition like the following is something which should be avoided:
+
+[source,java]
+----
+package com.example;
+
+import com.example.internal.Foo;
+
+public interface MyPublicService {
+
+    Foo doFoo();
+}
+----
+
+`MyPublicService` is part of the public API (as it is not part of an `internal` package),
+but the `doFoo()` method returns the internal type `Foo`.
+Users of `doFoo()` thus would have to deal with an implementation type, which is exactly what you wanted to prevent when splitting the API and implementation parts of the library.
+
+Unfortunately, inconsistent APIs like this are defined easily if one isn't very careful.
+This is where tooling is helpful: by searching for such malformed APIs in an automated way,
+they can be spotted early on and be avoided.
+
+== Introducing jQAssistant
+
+https://jqassistant.org/[jQAssistant] is an open-source tool allowing you to do exactly that.
+It parses a project's code base and creates a model of all the classes, methods, fields etc. in a https://neo4j.com/[Neo4j graph database].
+Using Neo4j's powerful https://neo4j.com/developer/cypher-query-language/[Cypher] query language, you then can execute queries to detect specific patterns and structures in your code base you are interested in.
```

Comment:
`specific patterns and structures in your code base you are interested in`

This sounds weird. Maybe remove `in your code base` or `specific patterns and structures you are interested in in your code base`? Even with the 2 `in`, it sounds more natural.